### ISSUE TITLE
Green alerts are preserved in etcd without ttl

### ISSUE THREAD

**Issue by anmolbabu** [Tendrl/notifier#83]
Consider the following:

1. A tendrl managed resource was down
2. An alert is generated for this with the severity as CRITICAL...
3. Now lets say the resource is back up..
4. Currently the alerting module detects the id of the CRITICAL alert for the resource and then modifies the same alert to indicate that the state has changed and now the alert carries a severity INFO(equivalent of OK in alerting module)...

The questions are:
1. Should the alert be rather deleted from etcd as its no longer relevant and we don't really need to show green(OK or INFO) alerts and let them only appear in Notification's list?
2. Should the alert be there with some ttl?

**Comment by anmolbabu** [comment]
@nthomas-redhat @r0h4n @shtripat @brainfunked Please provide your thoughts...

**Comment by shtripat** [comment]
I am favor of option-2 with ttl. This alert should rwmain in user's screen for sometime and then vanish.
Also if there is email notification about this green alert, that would inform the end user anyway.

**Comment by brainfunked** [comment]
I prefer option two: TTL. The TTL value must be set only after the alert is switched to GREEN. RED alerts must not have a TTL. Timestamps for all the state flips for an alert must be maintained.

However, one additional thing to consider. If an alert is for a state change from RED to GREEN, it's priority should be the same as the priority of the original alert. @anmolbabu what would this mean in terms of the alert processing in the alerting application?

**Comment by anmolbabu** [comment]
@brainfunked 
>I prefer option two: TTL. The TTL value must be set only after the alert is switched to GREEN. RED alerts must not have a TTL.

Ack. I think this can be easily added to current way of implementation need to practically verify it though.But yes the implementation should readily facilitate adding this...

>Timestamps for all the state flips for an alert must be maintained.
All state flip time stamps are not maintained currently.. However, the current and the immediate previous are maintained....
So, for example, 
1. A tendrl managed resource was down
2. An alert is generated for this with the severity as CRITICAL and with a timestamp...
3. Now lets say the resource is back up..
4. Now, the alerting module detects the id of the CRITICAL alert for the resource and then modifies the same alert to indicate that the state has changed and now the alert carries a severity INFO(equivalent of OK in alerting module)... and maintains the timestamp of clearing event in acket_at field of alert.

I feel if there is a CRITICAL Alert, and then a clearing alert this means the alert is no longer relevant and so we clear it and we will add ttl so that it stays there for some time.. But if there is again a CRITICAL alert, I think we should consider it as a new alert and it makes more sense according to me, specially when we set ttl for clearing alert and the alert vanishes from etcd after ttl expires...

>However, one additional thing to consider. If an alert is for a state change from RED to GREEN, it's priority should be the same as the priority of the original alert.
The current implementation has 2 fields and I tend to get confused b/w them So just to be sure which field exactly we are referring to , the 2 fields are:
1. Severity : This indicates the green-ness or red-ness of an alert.
2. Priority: Indicates the delivery guarantee level expected(currently this field exists but is not used as of now).. This field's value is dependent on the integration module that dispatches this alert as was discussed previously. Plus the current implementation is in accordance with what you said above(I felt it was a bug :) feels good to know its not a bug though ;) )

 @anmolbabu what would this mean in terms of the alert processing in the alerting application?
Please view my answer to this question above at the level of individual questions.

**Comment by brainfunked** [comment]
> I feel if there is a CRITICAL Alert, and then a clearing alert this means the alert is no longer relevant and so we clear it and we will add ttl so that it stays there for some time.. But if there is again a CRITICAL alert, I think we should consider it as a new alert and it makes more sense according to me, specially when we set ttl for clearing alert and the alert vanishes from etcd after ttl expires...

In most cases, it depends upon the actual alert condition. However, if an alert is critical enough to be marked CRITICAL, then it's resolution is equally critical. This is especially important based on the following:

> Priority: Indicates the delivery guarantee level expected(currently this field exists but is not used as of now).. This field's value is dependent on the integration module that dispatches this alert as was discussed previously. Plus the current implementation is in accordance with what you said above(I felt it was a bug :) feels good to know its not a bug though ;) )

I guess what I'm talking about is the alert priority. This, as you've correctly interpreted, deals with the delivery guarantees for the alert. Whereas,

> Severity : This indicates the green-ness or red-ness of an alert.

This is the "business logic" of the alert.

I think we're on the same track here.

**Comment by anmolbabu** [comment]
>>I feel if there is a CRITICAL Alert, and then a clearing alert this means the alert is no longer relevant and so we clear it and we will add ttl so that it stays there for some time.. But if there is again a CRITICAL alert, I think we should consider it as a new alert and it makes more sense according to me, specially when we set ttl for clearing alert and the alert vanishes from etcd after ttl expires...

>In most cases, it depends upon the actual alert condition. However, if an alert is critical enough to be marked CRITICAL, then it's resolution is equally critical.

Yes as I said, the priority of both alerts is high while the severity(an indicative of the red-ness or green-ness of the alert) is INFO for green alert and CRITICAL for a red alert..
cc: @brainfunked

**Comment by brainfunked** [comment]
> Yes as I said, the priority of both alerts is high while the severity(an indicative of the red-ness or green-ness of the alert) is INFO for green alert and CRITICAL for a red alert..

That seems alright then. All we need is for both alerts to be delivered with the same guarantees.

**Comment by anmolbabu** [comment]
>>Yes as I said, the priority of both alerts is high while the severity(an indicative of the red-ness or green-ness of the alert) is INFO for green alert and CRITICAL for a red alert..

>That seems alright then. All we need is for both alerts to be delivered with the same guarantees.

The current implementation already has this.
All that remains now is to add ttl then I'll send a PR for it.
cc: @brainfunked

**Comment by GowthamShanmugam** [comment]
@anmolbabu this issue is not required anymore, So please close this

**Comment by anmolbabu** [comment]
Closing this based on https://github.com/Tendrl/notifier/issues/83#issuecomment-330138319
I feel this is already fixed
Cc @GowthamShanmugam
